crystal_diffusion/train_mtp.py

```diff
@@ -0,0 +1,202 @@
+    """
+    structures = []
+    forces = []
+    with (open(filename, 'r') as f):
+        l_yaml = yaml.safe_load_all(f)
+        for d in l_yaml:  # loop over LAMMPS outputs and convert in pymatgen Structure objects
+            # lattice in yaml is 3 x 2 [0, x_lim]
+            # we assume a rectangular lattice for now with the 2nd coordinates as the lattice vectors
+            lattice = np.zeros((3, 3))
+            for i, x in enumerate(d['box']):
+                lattice[i, i] = x[1]
+            type_idx = d['keywords'].index('type')
+            species = [atom_dict[x[type_idx]] for x in d['data']]  # convert to atom type
+            coords_idx = [d['keywords'].index(x) for x in ['x', 'y', 'z']]
+            coords = [[x[i] for i in coords_idx] for x in d['data']]
+            pm_structure = Structure(lattice=lattice,
+                                     species=species,
+                                     coords=coords)
+            structures.append(pm_structure)
+            force_idx = [d['keywords'].index(x) for x in ['fx', 'fy', 'fz']]
+            structure_forces = [[x[i] for i in force_idx] for x in d['data']]
+            forces.append(structure_forces)
+    return structures, forces
+
+
+def extract_energy_from_thermo_log(filename: str) -> List[float]:
+    """Read energies from LAMMPS thermodynamic output file.
+
+    Args:
+        filename: path to LAMMPS thermodynamic output file in yaml format.
+
+    Returns:
+        list of energies (1 value per configuration)
+    """
+    with open(filename, 'r') as f:
+        log_yaml = yaml.safe_load(f)
+        kin_idx = log_yaml['keywords'].index('KinEng')
+        pot_idx = log_yaml['keywords'].index('PotEng')
+        energies = [x[kin_idx] + x[pot_idx] for x in log_yaml['data']]
+    return energies
+
+
+def prepare_mtp_inputs_from_lammps(output_yaml: List[str],
+                                   thermo_yaml: List[str],
+                                   atom_dict: Dict[int, Any]
+                                   ) -> Dict[str, Any]:
+    """Convert a list of LAMMPS output files and thermodynamic output files to MTP input format.
+
+    Args:
+        output_yaml: list of LAMMPS output files as yaml.
+        thermo_yaml: list of LAMMPS thermodynamic output files as yaml.
+        atom_dict: mapping of LAMMPS indices to atom type.
+
+    Returns:
+        dict with structure, energies and forces usable by MTP.
+    """
+    mtp_inputs = {
+        'structure': [],
+        'energy': [],
+        'forces': []
+    }
+    for filename in output_yaml:
+        structures, forces = extract_structure_and_forces_from_file(filename, atom_dict)
+        mtp_inputs['structure'] += structures
+        mtp_inputs['forces'] += forces
+    for filename in thermo_yaml:
+        mtp_inputs['energy'] += extract_energy_from_thermo_log(filename)
+    return mtp_inputs
+
+
+def train_mtp(train_inputs: Dict[str, Any], mlip_cmd_path: str, save_dir: str) -> MTPWithMLIP3:
```

Comment:
I think `train_inputs` should be a namedtuple https://docs.python.org/3/library/collections.html#collections.namedtuple.

That's a pretty lightweight way of "typing", ie telling the user exactly what is expected as input.

Reply:
changed